colibri/src/api/icons.rs

```diff
@@ -33,3 +57,117 @@ pub async fn get_icon(
+                                return query_icon_from_payload(state, payload, path)
+                                    .await
+                                    .into_response();
+                            }
+                            None | Some(false) => return StatusCode::OK.into_response(),
+                        }
+                    }
+
+                    // check when was the last time that the file got updated
+                    if let Ok(time) = metadata.modified() {
+                        if let Ok(duration) = SystemTime::now().duration_since(time) {
+                            if duration.as_secs() < MAX_ICON_RECHECK_PERIOD {
+                                // It was queried less than 12 hours ago so don't try again
+                                StatusCode::NOT_FOUND.into_response()
+                            } else {
+                                let _ = fs::remove_file(found_path).await;
+                                // Since we tried long ago enough retry again
+                                tokio::spawn(icons::query_icon_remotely(
+                                    payload.asset_id,
+                                    state.data_dir.clone(),
+                                    state.coingecko.clone(),
+                                ));
+                                StatusCode::ACCEPTED.into_response()
+                            }
+                        } else {
+                            error!("Error calculating elapsed duration");
+                            StatusCode::INTERNAL_SERVER_ERROR.into_response()
+                        }
+                    } else {
+                        // This shouldn't happen as we ship to platforms with metadata on files
+                        error!("Platform doesn't support last modified ts");
+                        StatusCode::INTERNAL_SERVER_ERROR.into_response()
+                    }
+                }
+                Err(e) => {
+                    error!("Failed to query icon for {} due to {}", payload.asset_id, e);
+                    StatusCode::NOT_FOUND.into_response()
+                }
+            }
+        }
+        None => {
+            // There is no reference locally to the file, query it. Ensure that if it is requested
+            // again only one task handles it.
+            query_icon_from_payload(state, payload, path)
+                .await
+                .into_response()
+        }
+    }
+}
+
+/// Helper function to update the status of the query in the shared state
+/// and start the query of an icon remotely.
+async fn query_icon_from_payload(
+    state: Arc<AppState>,
+    payload: AssetIconCheck,
+    path: std::path::PathBuf,
+) -> StatusCode {
+    let task_name = format!("{}_{}", QUERY_ICONS_TASK_PREFIX, payload.asset_id);
+    let mut tasks = state.active_tasks.lock().await;
+    if !tasks.insert(task_name.clone()) {
+        return StatusCode::ACCEPTED;
+    };
+    drop(tasks);
```

Comment:
this code here looks suspicious. Why get a mutable reference to the tasks, insert the new task and then put `tasks` out of scope with `drop`?

Reply:
This still looks suspiciously complicated. Can you at least add a small comment on the right of the drop `// BLA BLA` explaining why.

I don't have time to improve it unfortunately at this point.